src/wui/wui_api.c

```diff
@@ -0,0 +1,117 @@
+static struct fs_file api_file;
+static char _buffer[BDY_WUI_API_BUFFER_SIZE];
+
+static int char_streamer(const char *format, ...) {
+    int rv = 0;
+    va_list args;
+    va_start(args, format);
+    rv = vsnprintf(_buffer, BDY_WUI_API_BUFFER_SIZE, format, args);
+    va_end(args);
+    return rv;
+}
+
+static void wui_api_job(struct fs_file* file) {
+
+    const char* file_name = "test.gcode";
+    uint8_t sd_percent_done = (uint8_t)(webserver_marlin_vars_copy.sd_percent_done);
+    uint32_t print_duration = (uint32_t)(webserver_marlin_vars_copy.print_duration);
+
+    int response_len = char_streamer("{"
+            "\"file\":\"%s\","
+            "\"total_print_time\":%d, "
+            "\"progress\":{\"precent_done\":%d}"
+            "}",
+            file_name,
+            print_duration,
+            sd_percent_done
+            );
+    file->len = response_len;
+    file->data = (const char*)&_buffer;
+    file->index = response_len;
+    file->pextension = NULL;
+    file->flags = BDY_NO_FS_FLAGS;
+}
+
+static void wui_api_printer(struct fs_file* file) {
+
+    int32_t actual_nozzle = (int32_t) (webserver_marlin_vars_copy.temp_nozzle);
+    int32_t target_nozzle = (int32_t) (webserver_marlin_vars_copy.target_nozzle);
+    int32_t actual_heatbed = (int32_t) (webserver_marlin_vars_copy.temp_bed);
+    int32_t target_heatbed = (int32_t) (webserver_marlin_vars_copy.target_bed);
+
+    double x_pos_mm = (double) webserver_marlin_vars_copy.pos[X_AXIS_POS];
+    double y_pos_mm = (double) webserver_marlin_vars_copy.pos[Y_AXIS_POS];
+    double z_pos_mm = (double) webserver_marlin_vars_copy.pos[Z_AXIS_POS];
+    uint16_t print_speed = (uint16_t) (webserver_marlin_vars_copy.print_speed);
+    uint16_t flow_factor = (uint16_t) (webserver_marlin_vars_copy.flow_factor);
+    const char *filament_material = filaments[get_filament()].name;
+
+    int response_len = char_streamer(
+            "{\"temperature\":{"
+                    "\"tool0\":{\"actual\":%d, \"target\":%d},"
+                    "\"bed\":{\"actual\":%d, \"target\":%d}},"
+                    "\"xyz_pos_mm\":{"
+                    "\"x\":%.2f, \"y\":%.2f, \"z\":%.2f},"
+                    "\"print_settings\":{"
+                    "\"printing_speed\":%hd, \"flow_factor\":%hd, \"filament_material\":\"%s\"} }",
+
+            actual_nozzle, target_nozzle, actual_heatbed, target_heatbed,
+            x_pos_mm, y_pos_mm, z_pos_mm, print_speed, flow_factor,
+            filament_material);
+    file->len = response_len;
+    file->data = (const char*)&_buffer;
+    file->index = response_len;
+    file->pextension = NULL;
+    file->flags = BDY_NO_FS_FLAGS; // http server adds response header
+}
+
+struct fs_file* wui_api_main(char* uri,struct fs_file* file ) {
+
+    osStatus status = osMutexWait(wui_web_mutex_id, osWaitForever);
+    if (status == osOK) {
+        webserver_marlin_vars_copy = webserver_marlin_vars;
+    }
+    osMutexRelease(wui_web_mutex_id);
```

Comment:
Shouldn't this line be inside the if statement? You don't want to release if you haven't acquired it.

Reply:
shouldn't create an issue

Reply:
Really? At least it is misleading...